esmvaltool/diag_scripts/weighting/climwip.py

```diff
@@ -0,0 +1,438 @@
+"""Implementation of the climwip weighting scheme.
+
+Lukas Brunner et al. section 2.4
+https://iopscience.iop.org/article/10.1088/1748-9326/ab492f
+"""
+from collections import defaultdict
+import os
+import logging
+from datetime import datetime
+
+import yaml
+import matplotlib.pyplot as plt
+import numpy as np
+import seaborn as sns
+import xarray as xr
+from scipy.spatial.distance import pdist, squareform
+
+from esmvaltool.diag_scripts.shared import (ProvenanceLogger,
+                                            get_diagnostic_filename,
+                                            get_plot_filename, run_diagnostic,
+                                            select_metadata)
+
+logger = logging.getLogger(os.path.basename(__file__))
+
+
+def get_provenance_record(caption: str, ancestors: list):
```

Comment:
Nice idea to use type hints, maybe you could improve this a bit by using the `typing` module, so this would probably become something like:
```suggestion
def get_provenance_record(caption: str, ancestors: List[str]) -> Mapping[str, Union[str, List[str]]]:
```
? Or is that too complicated already? Keep in mind that less software engineering minded scientists also need to be able to understand the code.

I would also suggest using [mypy](https://mypy.readthedocs.io/en/latest/getting_started.html) to check that you actually got the types right, when I ran it, it found a few errors.

Reply:
I like to use typehints for documentation purposes and for myself to keep track of the types. I would probably want to keep it simple, and going all-in with `mypy` and `types` means a lot of fiddling which doesn't necessarily make it easier to grasp the code (and esmvaltool does not use mypy).

Reply:
Thanks for the suggestion of running mypy, it raised some easy to fix issues.